include/pybind11/detail/internals.h

```diff
@@ -602,13 +650,77 @@ inline local_internals &get_local_internals() {
     return *locals;
 }
 
+#if PYBIND11_INTERNALS_VERSION >= 6 && defined(Py_GIL_DISABLED)
+#    define PYBIND11_LOCK_INTERNALS(internals) std::unique_lock<std::mutex> lock((internals).mutex)
+#else
+#    define PYBIND11_LOCK_INTERNALS(internals)
+#endif
+
+template <typename F>
+inline auto with_internals(const F &cb) -> decltype(cb(get_internals())) {
+    auto &internals = get_internals();
+    PYBIND11_LOCK_INTERNALS(internals);
+    return cb(internals);
+}
+
+inline uint64_t splitmix64(uint64_t z) {
+    z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
+    z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
+    return z ^ (z >> 31);
+}
+
+template <typename F>
+inline auto with_instance_map(const void *ptr,
+                              const F &cb) -> decltype(cb(std::declval<instance_map &>())) {
+    auto &internals = get_internals();
+
+#if PYBIND11_INTERNALS_VERSION >= 6
+    // Hash address to compute shard, but ignore low bits. We'd like allocations
+    // from the same thread/core to map to the same shard and allocations from
+    // other threads/cores to map to other shards. Using the high bits is a good
+    // heuristic because memory allocators often have a per-thread
+    // arena/superblock/segment from which smaller allocations are served.
+    auto addr = reinterpret_cast<uintptr_t>(ptr);
+    uint64_t hash = splitmix64((uint64_t) (addr >> 20));
```

Comment:
If you don't hash the whole pointer, my suggestion would be to use a cheaper 32 bit hash.